src/mlpack/core/data/string_encoding_impl.hpp

```diff
@@ -71,23 +75,26 @@ void StringEncoding<EncodingPolicyType, DictionaryType>::CreateMap(
     const std::string& input,
     const TokenizerType& tokenizer)
 {
-  boost::string_view strView(input);
-  auto token = tokenizer(strView);
-
-  static_assert(
-      std::is_same<typename std::remove_reference<decltype(token)>::type,
-                   typename std::remove_reference<typename DictionaryType::
-                      TokenType>::type>::value,
-      "The dictionary token type doesn't match the return value type "
-      "of the tokenizer.");
-
-  // The loop below adds the extracted tokens to the dictionary.
-  while (!tokenizer.IsTokenEmpty(token))
+  for(int ngram = minNgram; ngram <= maxNgram; ngram++)
   {
-    if (!dictionary.HasToken(token))
-      dictionary.AddToken(std::move(token));
+    boost::string_view strView(input);
+    auto token = tokenizer(strView, ngram);
 
-    token = tokenizer(strView);
+    static_assert(
+        std::is_same<typename std::remove_reference<decltype(token)>::type,
+                     typename std::remove_reference<typename DictionaryType::
+                        TokenType>::type>::value,
+        "The dictionary token type doesn't match the return value type "
+        "of the tokenizer.");
```

Comment:
We should probably remove this check outside the loop, since it is costly to check multiple times, But I am not sure how